### DISCUSSION TITLE
Custom Delegating Handlers to support plug-ins

### DISCUSSION THREAD

**Discussion by dimevit** [dotnet/yarp#1935] [General] [answered]
Hi,

from the documentation I could not find out, if custom delegating handlers (http and websockets) like in ocelot are supported in YARP
ocelot docu for delegation handlers: https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html

I saw in some of the pull requet that such issues were requested and also implemented but I can not find the changes in the source code, probably they are not merged e.g. https://github.com/microsoft/reverse-proxy/pull/869

Thanks

**Comment by MihaZupan** [comment, marked as answer]
If you're using [direct forwarding](https://microsoft.github.io/reverse-proxy/articles/direct-forwarding.html), you control the `HttpClientHandler` instance yourself, so you can customize it however you wish.

If you're using the full YARP middleware (`.MapReverseProxy()`), you can inject custom delegating handlers by implementing a custom `IForwarderHttpClientFactory` and registering it in DI. See https://microsoft.github.io/reverse-proxy/articles/http-client-config.html

E.g.
```c#
services.AddSingleton<IForwarderHttpClientFactory, CustomHttpClientFactory>();

private sealed class CustomHttpClientFactory : ForwarderHttpClientFactory
{
    protected override WrapHandler(ForwarderHttpClientContext context, HttpMessageHandler handler)
    {
        return new MyHandler(handler);
    }
}
```

**Comment by Tratcher** [comment]
Is there a specific reason it needs to be delegating handlers? Yes those work in YARP as shown above, but you can also use aspnetcore middleware and YARP request/response transforms. All three have different strengths and weaknesses. What are you trying to accomplish?

**Reply by ChristinaBauer** [reply]
We evaluate to use YARP as API Gateway that exposes APIs. We need to add proprietary plugins (delegating handlers) that are either called for all routes (APIs) or for specific routes.

Example use cases:

- Protocol translation: expose websocket APIs via the API gateway and translate JSON messages to gRPC messages (this is an example to be called for specific routes, since not all microservices behind the API gateway will be gRPC based. Some will be websocket based as well)
- JSON transformation: Transform the JSON format of a websocket API call to a different JSON format
- Collect usage statistics (for all routes)
- Add proprietary logging
- Potentially add specific authentication/authorization (might not be necessary; we didn't look into YARP's capabilities yet)

We need bi-directional communication and hence go for websocket. Is it possible to call Middleware also for the direction server => client?

**Reply by Tratcher** [reply]
Middleware can intercept responses, but the mechanics are a bit trickier than those in delegating handlers.

**Reply by ChristinaBauer** [reply]
May I kindly ask you to contact me via mail? I wrote you a mail to Tratcher@outlook.com on November 28th, 2022. Thanks a lot!

**Reply by samsp-msft** [reply]
Some members of the team located in Europe should be reaching out as they are in a better position timezone wise.

As YARP is implemented as ASP.NET Core middleware, it can be incorporated with other middleware to handle situations like authentication/authorization and logging. In addition it uses routing which can be mixed with other endpoints, so for JSON transformation, and gRPC/websocket transformations, those are not really what YARP is designed for, but there is no reason why handlers can't be registered for those endpoints, that will then do what ever conversion is needed. Those can be mixed in the route table along with YARP routes for anything that doesn't need that kind of special handling, so that calling applications will see a unified url-space but each path can be handled differently.
